// @goatui/components/dist/esm/goat-breadcrumb-item.entry.js
import { r as registerInstance, h, H as Host } from './index-6bad8be4.js';

const breadcrumbItemCss = "*{box-sizing:border-box}:host{font-family:var(--font-family-base)}::selection{color:var(--color-white, #ffffff);background:var(--color-primary, var(--color-blue-60, #0f62fe))}::-webkit-scrollbar{width:var(--scrollbar-width, 0.75rem);height:var(--scrollbar-width, 0.75rem)}::-webkit-scrollbar-track{background:#f1f1f1}::-webkit-scrollbar-thumb{background:#888}::-webkit-scrollbar-thumb :hover{background:#555}.sr-only{position:absolute;width:1px;height:1px;padding:0;margin:-1px;overflow:hidden;clip:rect(0, 0, 0, 0);border:0}.center-content{display:flex;align-items:center;justify-content:center}:host([hidden]){display:none}@keyframes reveal{0%{opacity:0;transform:translateY(8px)}}";

const BreadcrumbItem = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.href = undefined;
    this.target = undefined;
    this.position = undefined;
    this.active = false;
  }
  render() {
    return (h(Host, { itemprop: 'itemListElement', itemscope: true, itemtype: 'http://schema.org/ListItem' }, this.active ? h("goat-text", null, h("span", { itemProp: 'name' }, h("slot", null)), h("meta", { itemProp: 'position', content: this.position })) : h("goat-link", { itemprop: 'item', href: this.href, target: this.target }, h("span", { itemProp: 'name' }, h("slot", null)), h("meta", { itemProp: 'position', content: this.position }))));
  }
};
BreadcrumbItem.style = breadcrumbItemCss;

export { BreadcrumbItem as goat_breadcrumb_item };
