// @goatui/components/dist/esm/goat-time-picker.entry.js
import { r as registerInstance, c as createEvent, h, H as Host, g as getElement } from './index-6bad8be4.js';
import { g as getComponentIndex } from './utils-02ac3efa.js';

const timePickerCss = "*{box-sizing:border-box}:host{font-family:var(--font-family-base)}::selection{color:var(--color-white, #ffffff);background:var(--color-primary, var(--color-blue-60, #0f62fe))}::-webkit-scrollbar{width:var(--scrollbar-width, 0.75rem);height:var(--scrollbar-width, 0.75rem)}::-webkit-scrollbar-track{background:#f1f1f1}::-webkit-scrollbar-thumb{background:#888}::-webkit-scrollbar-thumb :hover{background:#555}.sr-only{position:absolute;width:1px;height:1px;padding:0;margin:-1px;overflow:hidden;clip:rect(0, 0, 0, 0);border:0}.center-content{display:flex;align-items:center;justify-content:center}:host([hidden]){display:none}@keyframes reveal{0%{opacity:0;transform:translateY(8px)}}:host{display:block;margin-bottom:var(--spacing-3, 0.75rem);--input-border-radius:var(--border-radius, 3px);--input-border-style:solid;color:var(--text-primary, var(--color-gray-100, #161616))}.input-container{position:relative;display:flex;align-items:center;justify-content:center;background-color:var(--field-01, var(--color-gray-10, #f4f4f4));border:1px var(--input-border-style) var(--ui-04, var(--color-gray-50, #8d8d8d));border-radius:var(--input-border-radius);overflow:hidden;cursor:text;font-weight:var(--font-weight-regular);gap:var(--spacing-2, 0.5rem)}.input-container .input{flex:1;border:none;outline:none;background:none;width:100%;cursor:inherit;padding:0;color:var(--text-primary, var(--color-gray-100, #161616))}.input-container .input::placeholder{color:var(--text-placeholder, var(--color-gray-40, #a8a8a8))}.input-container .slot-container{display:flex;align-items:center;justify-content:center;line-height:0}.input-container .slot-container slot::slotted(*){padding-bottom:0 !important;margin-bottom:0 !important;color:var(--color-neutral-400, )}.input-container .slot-container slot::slotted(*) goat-icon{--color:var(--color-neutral-400, )}.input-container.start-slot-has-content .input{padding-left:var(--spacing-2, 0.5rem)}.input-container.end-slot-has-content .input{padding-right:var(--spacing-2, 0.5rem)}.input-container:not(.start-slot-has-content) .slot-container.start{display:none}.input-container:not(.end-slot-has-content) .slot-container.end{display:none}.input-container .input-action{cursor:pointer}.input-container .input-action:hover{--text-color:var(--text-secondary)}:host(.light) .input-container{background-color:var(--field-02, #ffffff)}.input-container.has-focus{outline:2px solid var(--color-primary, var(--color-blue-60, #0f62fe));border-color:transparent}:host-context(.input-state-success) .input-container{border-color:var(--support-success, #198038)}:host-context(.input-state-success) .input-container.has-focus{outline:2px solid var(--support-success, #198038)}:host-context(.input-state-warning) .input-container{border-color:var(--support-warning, #f1c21b)}:host-context(.input-state-warning) .input-container.has-focus{outline:2px solid var(--support-warning, #f1c21b)}:host-context(.input-state-error) .input-container{border-color:var(--support-error, #da1e28)}:host-context(.input-state-error) .input-container.has-focus{outline:2px solid var(--support-error, #da1e28)}:host([size=sm]) .input-container{padding:0.5rem 0.75rem;min-height:2.375rem}:host([size=sm]) .input-container .input{font-size:var(--text-sm-font-size, 0.875rem);line-height:var(--text-sm-line-height, 1.25rem);letter-spacing:var(--text-sm-letter-spacing, 0rem)}:host([size=sm]) .input-container slot::slotted(*){font-size:var(--text-sm-font-size, 0.875rem);line-height:var(--text-sm-line-height, 1.25rem);letter-spacing:var(--text-sm-letter-spacing, 0rem)}:host([size=md]) .input-container{padding:0.625rem 0.875rem;min-height:2.875rem}:host([size=md]) .input-container .input{font-size:var(--text-md-font-size, 1rem);line-height:var(--text-md-line-height, 1.5rem);letter-spacing:var(--text-md-letter-spacing, 0rem)}:host([size=md]) .input-container slot::slotted(*){font-size:var(--text-md-font-size, 1rem);line-height:var(--text-md-line-height, 1.5rem);letter-spacing:var(--text-md-letter-spacing, 0rem)}:host([size=lg]) .input-container{padding:0.75rem 1rem;min-height:3.375rem}:host([size=lg]) .input-container .input{font-size:var(--text-lg-font-size, 1.125rem);line-height:var(--text-lg-line-height, 1.75rem);letter-spacing:var(--text-lg-letter-spacing, 0rem)}:host([size=lg]) .input-container slot::slotted(*){font-size:var(--text-lg-font-size, 1.125rem);line-height:var(--text-lg-line-height, 1.75rem);letter-spacing:var(--text-lg-letter-spacing, 0rem)}:host([success]) .input-container{border-color:var(--color-success, #198038)}:host([error]) .input-container{border-color:var(--color-error, #da1e28);box-shadow:0 0 0 0.25rem rgba(239, 68, 68, 0.25)}:host([readonly]) .input-container{cursor:initial;background-color:var(--disabled-01, var(--color-gray-10, #f4f4f4)) !important;border-color:var(--disabled-01, var(--color-gray-10, #f4f4f4)) !important;opacity:1 !important}:host([disabled]){color:var(--text-disabled, var(--color-gray-30, #c6c6c6)) !important}:host([disabled]) .input-container{cursor:not-allowed;background-color:var(--disabled-01, var(--color-gray-10, #f4f4f4)) !important;border-color:var(--disabled-01, var(--color-gray-10, #f4f4f4)) !important;opacity:1 !important}:host([disabled]) .input-container .input{color:var(--text-disabled, var(--color-gray-30, #c6c6c6)) !important}input::-webkit-outer-spin-button,input::-webkit-inner-spin-button{-webkit-appearance:none;margin:0}input[type=number]{-moz-appearance:textfield}";

const TimePicker = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.goatInput = createEvent(this, "goat:input", 7);
    this.goatChange = createEvent(this, "goat:change", 7);
    this.goatBlur = createEvent(this, "goat:blur", 7);
    this.goatFocus = createEvent(this, "goat:focus", 7);
    this.gid = getComponentIndex();
    this.inputHandler = (ev) => {
      const input = ev.target;
      const oldValue = this.value;
      if (input) {
        this.value = input.value;
      }
      this.goatInput.emit(ev);
      if (oldValue !== this.value) {
        this.goatChange.emit(ev);
      }
    };
    this.keyDownHandler = (ev) => {
      if (ev.key === 'Escape') {
        this.clearInput(ev);
      }
    };
    this.blurHandler = (ev) => {
      this.hasFocus = false;
      this.goatBlur.emit(ev);
    };
    this.focusHandler = (ev) => {
      this.hasFocus = true;
      this.goatFocus.emit(ev);
    };
    this.clearInput = (evt) => {
      this.nativeInput.value = '';
      this.inputHandler(evt);
    };
    this.name = `goat-input-${this.gid}`;
    this.placeholder = undefined;
    this.size = 'md';
    this.disabled = false;
    this.readonly = false;
    this.value = '';
    this.configAria = {};
    this.hasFocus = false;
  }
  hasValue() {
    return this.getValue().length > 0;
  }
  getValue() {
    return (this.value || '').toString();
  }
  async getComponentId() {
    return this.gid;
  }
  /**
   * Sets focus on the native `input` in `goat-input`. Use this method instead of the global
   * `input.focus()`.
   */
  async setFocus() {
    if (this.nativeInput) {
      this.nativeInput.focus();
      this.hasFocus = true;
    }
  }
  /**
   * Sets blur on the native `input` in `goat-input`. Use this method instead of the global
   * `input.blur()`.
   */
  async setBlur() {
    if (this.nativeInput) {
      this.nativeInput.blur();
      this.hasFocus = false;
    }
  }
  componentWillLoad() {
    // If the ion-input has a tabindex attribute we get the value
    // and pass it down to the native input, then remove it from the
    // goat-input to avoid causing tabbing twice on the same element
    if (this.elm.hasAttribute('tabindex')) {
      const tabindex = this.elm.getAttribute('tabindex');
      this.tabindex = tabindex !== null ? tabindex : undefined;
      this.elm.removeAttribute('tabindex');
    }
    this.elm.getAttributeNames().forEach((name) => {
      if (name.includes('aria-')) {
        this.configAria[name] = this.elm.getAttribute(name);
        this.elm.removeAttribute(name);
      }
    });
  }
  render() {
    return (h(Host, { "has-focus": this.hasFocus, "has-value": this.hasValue() }, h("div", { class: {
        'input-container': true,
        'disabled': this.disabled,
        'has-focus': this.hasFocus,
      } }, h("input", { type: 'time', tabindex: this.tabindex, class: 'input input-native', disabled: this.disabled, readOnly: this.readonly, onKeyDown: this.keyDownHandler, onInput: this.inputHandler, onBlur: this.blurHandler, onFocus: this.focusHandler }))));
  }
  get elm() { return getElement(this); }
};
TimePicker.style = timePickerCss;

export { TimePicker as goat_time_picker };
